store/poi_rating.go

```diff
@@ -0,0 +1,147 @@
+package store
+
+import (
+	"context"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/bitmark-inc/autonomy-api/schema"
+	"github.com/bitmark-inc/autonomy-api/score"
+)
+
+// poi_rating.go is an extension of  poi.go
+// methods of interface are defined in poi.go
+func (m *mongoDB) GetPOIResourceMetric(poiID primitive.ObjectID) (schema.POIRatingsMetric, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	c := m.client.Database(m.database).Collection(schema.POICollection)
+	filter := bson.M{
+		"_id": poiID,
+	}
+	var result schema.POI
+	err := c.FindOne(ctx, filter).Decode(&result)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"prefix": mongoLogPrefix,
+			"poi ID": poiID.String(),
+			"error":  err,
+		}).Error("get poi fail")
+		return schema.POIRatingsMetric{}, err
+	}
+	return result.ResourceRatings, nil
+}
+
+func (m *mongoDB) UpdatePOIRatingMetric(accountNumber string, poiID primitive.ObjectID, ratings []schema.RatingResource) error {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	c := m.client.Database(m.database).Collection(schema.POICollection)
+
+	metric, err := m.GetPOIResourceMetric(poiID)
+
+	if err != nil {
+		log.WithFields(log.Fields{
+			"prefix": mongoLogPrefix,
+			"poi ID": poiID.String(),
+			"error":  err,
+		}).Error("get poi fail")
+		return err
+	}
+
+	profileMetric, err := m.GetProfilesRatingMetricByPOI(accountNumber, poiID)
+
+	if err != nil {
+		log.WithFields(log.Fields{
+			"prefix":        mongoLogPrefix,
+			"poi ID":        poiID.String(),
+			"acount number": accountNumber,
+			"error":         err,
+		}).Error("get poi fail")
+		return err
+	}
+
+	now := time.Now().UTC()
+	todayStartAt := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
+
+	resourceMap := make(map[string]schema.POIResourceRating)
+	for _, r := range metric.Resources { // make a current resources map
+		old, ok := resourceMap[r.Resource.ID]
+		if ok {
+			if old.LastUpdate < todayStartAt.Unix() {
+				r.LastDayScore = old.Score
+				r.LastDayRatings = old.Ratings
+			}
+		}
+		r.LastUpdate = now.Unix()
+		resourceMap[r.Resource.ID] = r
+	}
+
+	for _, r := range ratings { //add user rating (could be an empty one) to score
+		existPoiRating, ok := resourceMap[r.Resource.ID]
+		existProfileRating := schema.RatingResource{}
+		update := false
+		if !ok { // new
+			existPoiRating = schema.POIResourceRating{}
+			log.Info("New POI rating")
+		} else { // old , should be update
+			for _, pr := range profileMetric.Resources {
+				if pr.ID == r.ID {
+					existProfileRating = pr
+					update = true
+				}
+			}
+		}
```

Comment:
firstRating = (userA, resourceX, 5)
secondRating = (userB, resourceX, 3)

Now the second rating from user B will be handled like from user A. I suppose you just need to loop through `profileMetric.Resources` without the if clause.

Reply:
I have test on 3 cases :